### PR TITLE
Create symbols

### LINKED ISSUE #35: Define symbols for semantic checking

### PR DESCRIPTION
**Pull request by thinker227**
Closes #35. Implement super basic symbols for semantic analysis.

### CODE REVIEW

File: src/Draco.Compiler/Semantics/Symbol.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    /// <summary>
+    /// A variable symbol.
+    /// </summary>
+    public interface IVariable : ISymbol
+    {
+        /// <summary>
+        /// The type of the variable.
+        /// </summary>
+        IType Type { get; }
+        /// <summary>
```
Comment (LPeter1997) [inline src/Draco.Compiler/Semantics/Symbol.cs:30]: Please add a newline between these. If you know of the editorconfig rule, that would also be handly.

File: src/Draco.Compiler/Semantics/Symbol.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        /// The name of the symbol.
+        /// </summary>
+        string Name { get; }
+    }
+
+    /// <summary>
+    /// A variable symbol.
+    /// </summary>
+    public interface IVariable : ISymbol
```
Comment (LPeter1997) [inline src/Draco.Compiler/Semantics/Symbol.cs:24]: The name of this symbol?

File: src/Draco.Compiler/Semantics/Symbol.cs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        /// <summary>
+        /// The kind of variable.
+        /// </summary>
+        VariableKind Kind { get; }
+    }
+
+    /// <summary>
+    /// A parameter symbol.
+    /// </summary>
+    public interface IParameter : IVariable
```
Comment (LPeter1997) [inline src/Draco.Compiler/Semantics/Symbol.cs:39]: The name of this symbol?

File: src/Draco.Compiler/Semantics/Symbol.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        /// <summary>
+        /// The function the parameter is a parameter to.
+        /// </summary>
+        IFunction Function { get; }
+    }
+
+    /// <summary>
+    /// A function symbol.
+    /// </summary>
+    public interface IFunction : ISymbol
```
Comment (LPeter1997) [inline src/Draco.Compiler/Semantics/Symbol.cs:50]: The name of this symbol?

File: src/Draco.Compiler/Semantics/Symbol.cs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    /// </summary>
+    public interface IType : ISymbol
+    {
+        // This currently doesn't do anything.
+    }
+
+    /// <summary>
+    /// The kind of a variable.
+    /// </summary>
+    public enum VariableKind
```
Comment (LPeter1997) [inline src/Draco.Compiler/Semantics/Symbol.cs:73]: Maybe name it `Mutability` as variables will also need to store if they are local or global.

**Review by LPeter1997 — CHANGES_REQUESTED** [review]
Thanks, just a few things.

File: src/Draco.Compiler/Semantics/Symbol.cs (hunk 2 above)
Comment (LPeter1997) [reply src/Draco.Compiler/Semantics/Symbol.cs:24]: Ah missed it, thanks

Comment (LPeter1997) [reply src/Draco.Compiler/Semantics/Symbol.cs:24]: That raises a question if we will have unnamed symbol, like the unreachable type or types resulting from complex type expressions like `[]int32`

Comment (thinker227) [reply src/Draco.Compiler/Semantics/Symbol.cs:24]: Unreachable could just have an unspeakable name (ex. `<>_unreachable`) and its own symbol (ex. `IUnreachableType`), and array types could also have their own symbol and literally just the name `[]T`. Question is what the point of the names actually is, for unreachable an unspeakable name works just fine and for for arrays they're a constructed type anyway, it's also kind of unspeakable.

Comment (LPeter1997) [reply src/Draco.Compiler/Semantics/Symbol.cs:24]: Alright, let's go with this design for now

**Review by LPeter1997 — APPROVED** [review]
Looks good to me. Please retarget to a new branch like `semantic-checking` or even just `symbol-resolution`

**Pull request merged**